curvefs/src/mds/topology/topology.cpp

```diff
@@ -1427,122 +1468,129 @@ TopoStatusCode TopologyImpl::GenInitialCopysetAddrBatchForPool(
-                    replicaMap.emplace(it.first, replicaNum);
-                } else {
-                    LOG(ERROR) << "get pool failed when choose metaservers,"
-                               << " the poolId = " << it.first;
-                    return TopoStatusCode::TOPO_POOL_NOT_FOUND;
+    std::vector<Pool> poolList;
+    for (const auto &it : poolMap_) {
+        if (GetMetaServerNumInPool(it.first, filter) != 0) {
+            poolList.push_back(it.second);
+        }
+    }
+
+    // sort pool list by copyset average num
+    std::sort(poolList.begin(), poolList.end(),
+        [=](const Pool& a, const Pool& b) {
+            PoolIdType poolId1 = a.GetId();
+            PoolIdType poolId2 = b.GetId();
+            uint32_t copysetNum1 = GetCopySetNumInPool(poolId1);
+            uint32_t copysetNum2 = GetCopySetNumInPool(poolId2);
+            uint32_t metaserverNum1 = GetMetaServerNumInPool(poolId1, filter);
+            uint32_t metaserverNum2 = GetMetaServerNumInPool(poolId2, filter);
+            double avaCopysetNum1 = copysetNum1 * 1.0 / metaserverNum1;
+            double avaCopysetNum2 = copysetNum2 * 1.0 / metaserverNum2;
+            return avaCopysetNum1 < avaCopysetNum2;
+        });
+
+    while (copysetList->size() < needCreateNum) {
+        for (auto it = poolList.begin(); it != poolList.end();) {
+            PoolIdType poolId = it->GetId();
+            uint16_t replicaNum = it->GetReplicaNum();
+            std::list<CopysetCreateInfo> tempCopysetList;
+            TopoStatusCode ret = GenCopysetAddrBatchForPool(poolId,
+                                    replicaNum, &tempCopysetList);
+            if (TopoStatusCode::TOPO_OK == ret) {
+                copysetList->splice(copysetList->end(), tempCopysetList);
+                if (copysetList->size() >= needCreateNum) {
+                    return TopoStatusCode::TOPO_OK;
                 }
+                it++;
             } else {
-                replicaNum = iter->second;
-            }
-            // find the available metaserver group
-            if (it.second.size() == replicaNum) {
-                for (const auto &item : it.second) {
-                    metaServers->emplace(item.second.GetId());
-                }
-                *poolId = it.first;
-                return TopoStatusCode::TOPO_OK;
+                LOG(WARNING) << "Generate " << needCreateNum
+                        << " copyset addr for pool " << poolId
+                        << "fail, statusCode = " << TopoStatusCode_Name(ret);
+                it = poolList.erase(it);
             }
         }
+
+        if (copysetList->size() == 0) {
+            LOG(ERROR) << "can not find available metaserver for copyset.";
+            return TopoStatusCode::TOPO_METASERVER_NOT_FOUND;
+        }
+    }
+
+    return TopoStatusCode::TOPO_OK;
+}
+
+// gen copyset addr
+// 1. Gen addr on the pool without copyset, if the number of copy is enough,
```

Comment:
What does the addr mean? May be can give an example.

Reply:
done